adapters/repos/db/queue/queue.go

```diff
@@ -0,0 +1,1046 @@
+
+type DiskQueue struct {
+	// Logger for the queue. Wrappers of this queue should use this logger.
+	Logger           logrus.FieldLogger
+	staleTimeout     time.Duration
+	taskDecoder      TaskDecoder
+	scheduler        *Scheduler
+	id               string
+	dir              string
+	onBatchProcessed func()
+	metrics          *Metrics
+
+	// m protects the disk operations
+	m            sync.RWMutex
+	lastPushTime time.Time
+	closed       bool
+	w            *chunkWriter
+	r            *chunkReader
+	recordCount  uint64
+	diskUsage    int64
+
+	rmLock sync.Mutex
+}
+
+type DiskQueueOptions struct {
+	// Required
+	ID          string
+	Scheduler   *Scheduler
+	Dir         string
+	TaskDecoder TaskDecoder
+
+	// Optional
+	Logger           logrus.FieldLogger
+	StaleTimeout     time.Duration
+	ChunkSize        uint64
+	OnBatchProcessed func()
+	Metrics          *Metrics
+}
+
+func NewDiskQueue(opt DiskQueueOptions) (*DiskQueue, error) {
+	if opt.ID == "" {
+		return nil, errors.New("id is required")
+	}
+	if opt.Scheduler == nil {
+		return nil, errors.New("scheduler is required")
+	}
+	if opt.Dir == "" {
+		return nil, errors.New("dir is required")
+	}
+	if opt.TaskDecoder == nil {
+		return nil, errors.New("task decoder is required")
+	}
+	if opt.Logger == nil {
+		opt.Logger = logrus.New()
+	}
+	opt.Logger = opt.Logger.WithField("queue_id", opt.ID)
+
+	if opt.Metrics == nil {
+		opt.Metrics = NewMetrics(opt.Logger, nil, nil)
+	}
+	if opt.StaleTimeout <= 0 {
+		opt.StaleTimeout = 5 * time.Second
+	}
+	if opt.ChunkSize <= 0 {
+		opt.ChunkSize = defaultChunkSize
+	}
+
+	// create the directory if it doesn't exist
+	err := os.MkdirAll(opt.Dir, 0o755)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create directory")
+	}
+
+	// create chunk reader
+	r, err := newChunkReader(opt.Dir)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create chunk reader")
+	}
+
+	// create chunk writer
+	w, err := newChunkWriter(opt.Dir, r, opt.Logger, opt.ChunkSize)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create chunk writer")
+	}
+
+	q := DiskQueue{
+		id:               opt.ID,
+		scheduler:        opt.Scheduler,
+		dir:              opt.Dir,
+		Logger:           opt.Logger,
+		staleTimeout:     opt.StaleTimeout,
+		taskDecoder:      opt.TaskDecoder,
+		metrics:          opt.Metrics,
+		onBatchProcessed: opt.OnBatchProcessed,
+		r:                r,
+		w:                w,
+	}
+
+	// determine the number of records stored on disk
+	// and the disk usage
+	err = q.analyzeDisk()
+	if err != nil {
+		return nil, err
+	}
+
+	// set the last push time to now
+	q.lastPushTime = time.Now()
+
+	return &q, nil
+}
+
+// Close the queue, prevent further pushes and unregister it from the scheduler.
+func (q *DiskQueue) Close() error {
+	if q == nil {
+		return nil
+	}
+
+	q.m.Lock()
+	if q.closed {
+		q.m.Unlock()
+		return errors.New("queue already closed")
+	}
+	q.closed = true
+	q.m.Unlock()
+
+	q.scheduler.UnregisterQueue(q.id)
```

Comment:
I am trying to understand why we do the q.m.Unlock() just to unregister the queue and then lock again. I think it is only because the UnregisterQueue uses the Pause which in turn locks too, leading to a deadlock if it is already locked. I think in this path then we have three times a lock on the same mutex. Should be more efficient to keep it locked through the complete block, right? I do not think this is a hot path and I could be totally wrong so feel free to ignore. I have seen in some places that we do a nonblocking version of the function to avoid this, in this case we could have some PauseNoLock method that pauses the queue without explicitly locking since it assumes it is already locked. I feel like this could also be more error pruned but more efficient. Could you tell me what are your thoughts on this so I can try to be consistent later?

Reply:
This function is `Close`, so it's usually called only during the node shutdown, or when we drop a shard. So basically it's not a hot path, as long as it's functionally correct and easy to read (I hope), anything is fine.
Other important point: it's usually called only once, not concurrently.

(Then why lock if it's gonna be called once? It's to prevent Push if the queue is closed)

Then about locking: We first lock the `q.closed` variable so we can prevent concurrent access to the rest of the function. Only one goroutine can skip the if statement l.175, all the other ones return l.177.
So we know that we can unregister only once. (I could have used a `sync.Once` for example, it would have worked the same)

> I think in this path then we have three times a lock on the same mutex.

Not exactly: `UnregisterQueue` is not a method of the queue, it's a method of the `Scheduler`.

Because it's a completely different entity (there can be N queues, but only 1 scheduler), we cannot lock while we are asking the scheduler to deregister the queue: the scheduler could be calling the queue at the same time (for chunk deletion for example), which would lead to a deadlock.

Then, only when we know the queue is unregistered from the scheduler (i.e. the scheduler is guaranteed NOT to read anything from the queue) we can close the files.

Basically, even if it's one single function, it's 3 different steps:
- setting the `q.closed` to true, but only once, filtering out any other call to Close
- unregistering the queue: can take a while if the scheduler is busy
- closing the resources